ext/opentelemetry-ext-asgi/src/opentelemetry/ext/asgi/__init__.py

```diff
@@ -0,0 +1,206 @@
+            return StatusCanonicalCode.DEADLINE_EXCEEDED
+        return StatusCanonicalCode.INTERNAL
+    return StatusCanonicalCode.UNKNOWN
+
+
+def collect_request_attributes(scope):
+    """Collects HTTP request attributes from the ASGI scope and returns a
+    dictionary to be used as span creation attributes."""
+
+    port = scope.get("server")[1]
+    server_host = scope.get("server")[0] + (
+        ":" + str(port) if port != 80 else ""
+    )
+    http_url = scope.get("scheme") + "://" + server_host + scope.get("path")
+    if scope.get("query_string"):
+        http_url = http_url + ("?" + scope.get("query_string").decode("utf8"))
+
+    result = {
+        "component": scope.get("type"),
+        "http.method": scope.get("method"),
+        "http.server_name": scope.get("server")[0],
+        "http.scheme": scope.get("scheme"),
+        "http.host": server_host,
+        "host.port": port,
+        "http.flavor": scope.get("http_version"),
+        "http.target": scope.get("path"),
+        "http.url": http_url,
+    }
+
+    if "client" in scope:
+        result["net.peer.ip"] = scope.get("client")[0]
+        result["net.peer.port"] = scope.get("client")[1]
+
+    return result
+
+
+def set_status_code(span, status_code):
+    """Adds HTTP response attributes to span using the status_code argument."""
+    try:
+        status_code = int(status_code)
+    except ValueError:
+        span.set_status(
+            Status(
+                StatusCanonicalCode.UNKNOWN,
+                "Non-integer HTTP status: " + repr(status_code),
+            )
+        )
+    else:
+        span.set_attribute("http.status_code", status_code)
+        span.set_status(Status(http_status_to_canonical_code(status_code)))
+
+
+def get_default_span_name(scope):
+    """Calculates a (generic) span name for an incoming HTTP request based on the ASGI scope."""
+
+    # TODO: Update once
+    #  https://github.com/open-telemetry/opentelemetry-specification/issues/270
+    #  is resolved
+    return scope.get("path", "/")
+
+
+class OpenTelemetryMiddleware:
+    """The ASGI application middleware.
+
+    This class is an ASGI middleware that starts and annotates spans for any
+    requests it is invoked with.
+
+    Args:
+        app: The ASGI application callable to forward requests to.
+    """
+
+    def __init__(self, app):
+        self.app = guarantee_single_callable(app)
+        self.tracer = trace.tracer_source().get_tracer(__name__, __version__)
+
+    async def __call__(self, scope, receive, send):
+        """The ASGI application
+
+        Args:
+            scope: A ASGI environment.
+            receive: An awaitable callable yielding dictionaries
+            send: An awaitable callable taking a single dictionary as argument.
+        """
+
+        parent_span = propagators.extract(get_header_from_scope, scope)
+        span_name = get_default_span_name(scope)
+
+        with self.tracer.start_as_current_span(
```

Comment:
If i'm reading this corrrectly, we'll have three spans that are emitted for every request:

1. a span when the function is called.
2. a span when the receive is called
3. a span when the send is called.

In the wsgi implementation, we just have a single span that represents the whole request. For most services, I think this makes sense, as it enable a single http request to map to a single span.

This scenario feels like it would cause three spans to be emitted per http request. Although it provides more fidelity, it can cause a lot of noise on a tracing backend, and I worry that the noise may outweigh the value of being this granular.

If the purpose of getting these traces are timings, could we just add span attributes to help someone understand the timings?

Probably need a second or third opinion on this, just throwing my two cents in.

Reply:
You are indeed reading it correctly. 3 spans will be emitted for every single http request.

And while I can understand your arguments against having 3 spans for http requests, here is the rationale for why the code is as it is, and why it is different from the WSGI implementation. 

* ASGI supports more than http, for example websockets. A single connection span for websockets would hide every single message send back and forth between the client and the server, while emitting the connection span and simply having the messages would omit the context of the messages.
* ASGI supports HTTP2, and amongst other things server push, which I believe also calls for the inclusion of a specialized send span. 

If you want to avoid these extra spans, we can change the code to detect simple http connections and not emit them on these requests, thus only emitting the connection span.
- Potentially this could be a configurable option too.

Reply:
+1 for multiple spans. In the past I've instrumented WSGI applications to include spans for reading from `wsgi.input` in order to have the time spent reading from the client explicitly called out. I agree it can be a bit noisy, but it's nice to be able to isolate time spent in IO with the client from everything else. Not sure how accessible that data would be as span attributes.

Reply:
I'm willing to be amenable to see how this turns out, as I'm not very familiar with asgi behavior in the wild. It definitely make sense for websockets.

I think at worst we'll just have to add a configuration variable to enable / disable these, and it's a strong argument that with it being async, this type of span will matter much more.

Reply:
So the way I hear this @toumorokoshi is that we go with multiple spans for now, and then add a configuration variable later if people seem to want it? - Aka. we add it in a later PR, if the need arises?

Reply:
Yep! That's what I was thinking.

Reply:
Resolving this conversation, feel free to unresolve if you do not feel like the conversation has been resolved.